File: Frontend/src/app/model/interface/note.ts
import {User} from './user';

/*
export interface Note {
  title: string;
  categoryId: string;
  tagIds: string[];
  status: string;
  content: string;
}

 */

export interface Category {
  id: string;
  name: string;
  noteCount: number;
}

export interface Tag{
  id: string;
  name: string;
  noteCount: number;
}

export enum noteStatus{
  ACTIVE = "ACTIVE",
  ARCHIVED = "ARCHIVED"
}

export interface noteCreated{
  id: string;
  title: string;
  content: string;
  author: User;
  category: Category;
  tagIds: Tag[];
  createdAt: Date;
  updatedAt: Date;
  status: noteStatus;
}

export interface Note {
  id: string;
  title: string;
  content: string;
  author?: {
    id: string;
    name: string;
  };
  category: Category;
  tags: Tag[];
  createdAt: string;
  updatedAt: string;
  status?: noteStatus;
}

export interface CreateNoteRequest {
  title: string;
  content: string;
  categoryId: string;
  tagIds: string[];
  status: noteStatus;
}

File: Backend/src/main/java/com/notes/Backend/domain/entities/Note.java
package com.notes.Backend.domain.entities;

import com.notes.Backend.domain.NoteStatus;
import jakarta.persistence.*;
import lombok.*;

import java.time.LocalDateTime;
import java.util.HashSet;
import java.util.Objects;
import java.util.Set;
import java.util.UUID;

@Entity
@Table(name = "notes")
@NoArgsConstructor
@AllArgsConstructor
@Getter
@Setter
@Builder
public class Note {

    @Id
    @GeneratedValue(strategy = GenerationType.UUID)
    private UUID id;

    @Column(nullable = false)
    private String title;

    @Column(nullable = false, columnDefinition = "TEXT")
    private String content;

    @Column(nullable = false)
    @Enumerated(EnumType.STRING)
    private NoteStatus status;

    @Column(nullable = false)
    private LocalDateTime createdAt;

    @Column(nullable = false)
    private LocalDateTime updatedAt;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "author_id", nullable = false)
    private User author;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "category_id", nullable = false)
    private Category category;

    @ManyToMany
    @JoinTable(
            name = "note_tags",
            joinColumns = @JoinColumn(name = "note_id"),
            inverseJoinColumns = @JoinColumn(name = "tag_id")
    )
    private Set<Tag> tags = new HashSet<>();

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        Note note = (Note) o;
        return Objects.equals(id, note.id) && Objects.equals(title, note.title)
                && Objects.equals(content, note.content) && status == note.status
                && Objects.equals(createdAt, note.createdAt)
                && Objects.equals(updatedAt, note.updatedAt);
    }

    @Override
    public int hashCode() {
        return Objects.hash(id, title, content, status, createdAt, updatedAt);
    }

    @PrePersist
    protected void onCreate(){
        LocalDateTime now = LocalDateTime.now();
        this.createdAt = now;
        this.updatedAt = now;
    }

    @PreUpdate
    protected void onUpdate(){
        this.updatedAt = LocalDateTime.now();
    }
}
